src/chain/candidate.rs

```diff
@@ -0,0 +1,141 @@
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::OnlinePayload;
+use crate::{id::PublicId, utils::LogIdent, utils::XorTargetInterval};
+use log::LogLevel;
+use std::collections::BTreeSet;
+
+/// A candidate (if any) may be in different stages of the resource proof process.
+/// When we consensus to accept them for resource proof, move to `AcceptedForResourceProof`
+/// with the value all elder expect to get from `CandidateInfo`.
+/// If we consensus to refuse them: reset to None.
+/// If we consensus to accept them: move to ApprovedWaitingSectionInfo until we are available to
+/// accept a new candidate, at which point reset to None.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
+pub enum Candidate {
+    /// No-one is currently in the resource proof process.
+    /// We can take on a new candidate on consensus of an `ExpectCandidate` event.
+    None,
+    /// We accepted a candidate to perform resource proof in this section. We are waiting for
+    /// them to send their `CandidateInfo` before starting the actual resource proof.
+    AcceptedForResourceProof {
+        target_interval: XorTargetInterval,
+        old_public_id: PublicId,
+    },
+    /// We consensused the candidate online. We are waiting for the SectionInfo to consensus
+    /// and this new node to start handling events before allowing a new candidate.
+    ApprovedWaitingSectionInfo { new_pub_id: PublicId },
+}
+
+impl Candidate {
+    /// Return true if already has a candidate
+    pub fn has_resource_proof_candidate(&self) -> bool {
+        *self != Candidate::None
+    }
+
+    /// Forget about the current candidate.
+    pub fn reset_candidate(&mut self) {
+        *self = Candidate::None;
+    }
+
+    /// Forget about the current candidate if it is a member of the given section.
+    pub fn reset_candidate_if_member_of(&mut self, members: &BTreeSet<PublicId>) {
+        if let Candidate::ApprovedWaitingSectionInfo { ref new_pub_id } = self {
+            if members.contains(new_pub_id) {
+                *self = Candidate::None;
+            }
+        }
+    }
+
+    /// Our section decided that the candidate should be selected next.
+    /// Pre-condition: !has_resource_proof_candidate.
+    pub fn accept_as_candidate(
+        &mut self,
+        old_public_id: PublicId,
+        target_interval: XorTargetInterval,
+    ) {
+        if self.has_resource_proof_candidate() {
+            log_or_panic!(
+                LogLevel::Error,
+                "accept_as_candidate when processing one already"
+            );
+        }
+
+        *self = Candidate::AcceptedForResourceProof {
+            old_public_id: old_public_id,
+            target_interval: target_interval,
+        };
+    }
+
+    /// Return the target interval if we are waiting for candidate info for that PublicId.
+    pub fn waiting_for_candidate_interval(
```

Comment:
If you agree with https://github.com/maidsafe/routing/pull/1648/files#r288895197, then maybe also rename this to `accepted_target_interval` (or similar).

Reply:
Will go for `accepted_candidate_target_interval`.